content/en/BestPractices/akhan_input_validation.md

```diff
@@ -0,0 +1,155 @@
+does *not* check if the quantity is positive.
+
+This small mistake is costly. Users can freely input negative values for
+`quantity`. If an attacker adds `-100` to their basket, the Juice Shop adds 100
+times the value of the item to their wallet.
+
+### Impact
+
+Without such checks, this error could lead to substantial business losses.
+
+### Prevention
+
+In the following code sample, we include code that:
+
+- Checks if the quantity is greater than zero (`quantity > 0`)
+- Runs a boolean check for decimal integers (`Number.isInteger(quantity, 10)`)
+- Returns an HTTP 400 error if the noted checks fail
+
+```ts
+async function quantityCheck (req: Request, res: Response, next: NextFunction, id: number, quantity: number) {
+  const product = await QuantityModel.findOne({ where: { ProductId: id } })
+  if (!product) {
+    throw new Error('No such product found!')
+  }
+
+ // Check if the quantity is a positive decimal integer value, if not send a `400` error response
+ if(!Number.isInteger(quantity, 10) || quantity <= 0) {
+      res.status(400).json({ error: res.__('Invalid Quantity of items.') })
+ }
+
+  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
+    if (product.quantity >= quantity) { // enough in stock?
+      next()
+    } else {
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+## Best Practices
+
+- Assume that any input can be an attack
+- Validate all user input before processing it
+  - The [OWASP Input Validation Cheat Sheet
+    recommends](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#goals-of-input-validation)
+that you use input validation as early as possible in the data flow,
+- Include an **allow list** of valid inputs. Attackers can often bypass the deny
+  lists. For example, you could use the following to validate a US zip code:
+
+```ts
+// define the allow list regex for US Zip code
+const validZipPattern = /^\d{5}(-\d{4})?$/;
+// check if the user input matches our defined allow list regex
+if (!validZipPattern.test(userInput) {
+ 		throw new Error("InvalidZipCode");
+}	
+```
+
+- Log all input validation failures. The information you get can help you detect
+  intrusions, and identify attackers who try to send invalid inputs. For this
+code sample, you can add the following logging statement: `logger.warn("Invalid zip code provided.")`:
+
+```ts
+// define the allow list regex for US Zip code
+const validZipPattern = /^\d{5}(-\d{4})?$/;
+// check if the user input matches our defined allow list regex
+if (!validZipPattern.test(userInput) {
+  // log the input validation failure
+  logger.warn("Invalid zip code provided.");
+  throw new Error("InvalidZipCode");
+}
+```
```

Comment:
Revise formatting per Markdown requirements

````suggestion
that you use input validation as early as possible in the data flow,
- Include an **allow list** of valid inputs. Attackers can often bypass the deny
  lists. For example, you could use the following to validate a US zip code:

  ```ts
  // define the allow list regex for US Zip code
  const validZipPattern = /^\d{5}(-\d{4})?$/;
  // check if the user input matches our defined allow list regex
  if (!validZipPattern.test(userInput) {
   		throw new Error("InvalidZipCode");
  }	
  ```

- Log all input validation failures. The information you get can help you detect
  intrusions, and identify attackers who try to send invalid inputs. For this
  code sample, you can add the following logging statement: `logger.warn("Invalid zip code provided.")`:

  ```ts
  // define the allow list regex for US Zip code
  const validZipPattern = /^\d{5}(-\d{4})?$/;
  // check if the user input matches our defined allow list regex
  if (!validZipPattern.test(userInput) {
    // log the input validation failure
    logger.warn("Invalid zip code provided.");
    throw new Error("InvalidZipCode");
  }
  ```
````